Introduce an open file aware cache cleanup policy

christge-sft opened the pull request:
Upgrade cache cleanup by adopting an open file aware LRU policy.

A Posix Cache Manager holds the information of refcounting, which is passed to a Posix Quota Manager.
During cache cleanup, if a file that would otherwise be selected for cleanup is open will be skipped, since it can not be removed from the cache without breaking the process that holds it open.

There are unit and integration tests for the features.

christge-sft committed:
[CORE] create backend for open file aware LRU $ cleanup

christge-sft committed:
[INTEGRATION] get the information of open files from the Cache Manager

christge-sft committed:
[TESTING] add logic to the cleanuplru integration test

vvolkl commented on test/unittests/t_smallhash.cc:
```diff
@@ -5,28 +5,32 @@
-  return static_cast<uint32_t>(*(reinterpret_cast<const uint32_t *>(key.digest) + 1));
+  static uint32_t hasher_md5(const shash::Md5 &key) {
+    // Don't start with the first bytes, because == is using them as well
+    return (uint32_t) * (reinterpret_cast<const uint32_t *>(key.digest) + 1);
```
This reverts a fix for clang-format, let's keep the static cast

vvolkl commented on cvmfs/mountpoint.cc:
```diff
@@ -308,6 +308,13 @@ FileSystem::PosixCacheSettings FileSystem::DeterminePosixCacheSettings(
     settings.do_refcount = false;
   }
 
+  if (options_mgr_->GetValue(MkCacheParm("CVMFS_CACHE_OF_AWARE_CLEANUP", instance),
```
The abbreviation OF isn't very clear in this name. What about something like `CVMFS_CACHE_NONOPENLRU`  or `CVMFS_CACHE_CLEANUP_NONOPENLRU`?

christge-sft replied:
As discussed in person, I am totally in favor of this change. I would go with the last (longer but more self explanatory) version.

Q: I use this "of" abbreviation for local variables related to the feature. Is it an issue there? (they are not part of an interface, they're either function local variables, or private/protected class members).

vvolkl replied:
Not a problem if they're not part of the interface.

vvolkl commented on test/src/034-cachecleanup/main:
```diff
@@ -2,21 +2,132 @@
 
 cvmfs_test_name="Cache cleanup"
 
```
Can you add a short "docstring" style comments to the test, summarizing what it is intending to test?

vvolkl commented on test/src/034-cachecleanup/main:
```diff
@@ -2,21 +2,132 @@
+    cat $file > /dev/null
+  done < $FILE_DB
+
+  # Keep the first half filse open
```
```suggestion
  # Keep the first half of files open
```
typo

vvolkl commented on test/src/034-cachecleanup/main:
```diff
@@ -2,21 +2,132 @@
+  sudo cvmfs_talk -i atlas.cern.ch cleanup $CLEANUP_TARGET || $(echo "FAILED" && return 31)
+  local cache_size_after_cleanup=$(get_cache_size_quiet)
+
+  #Check that cache size has been reduced at first and then that the goal has been reached
```
```suggestion
  # Check that cache size has been reduced at first and then that the goal has been reached
```

vvolkl commented on test/src/034-cachecleanup/main:
```diff
@@ -2,21 +2,132 @@
+  sudo cvmfs_talk -i atlas.cern.ch cleanup $CLEANUP_TARGET || $(echo "FAILED" && return 41)
+  local cache_size_after_cleanup=$(get_cache_size_quiet)
+
+  #Check that cache size has been reduced at first and then that the goal has been reached
```
```suggestion
  # Check that cache size has been reduced at first and then that the goal has been reached
```

vvolkl commented on cvmfs/fd_refcount_mgr.cc:
```diff
@@ -85,6 +85,12 @@ FdRefcountMgr *FdRefcountMgr::Clone() {
 }
 
+bool FdRefcountMgr::Contains(const shash::Any& hash){
+  int result{-1};
```
MacOS doesn't like this declaration:
```
2025-07-04T13:19:01.5144270Z /Users/runner/work/cvmfs/cvmfs/cvmfs/fd_refcount_mgr.cc:89:13: error: expected ';' at end of declaration
2025-07-04T13:19:01.5246470Z   int result{-1};
2025-07-04T13:19:01.5347830Z             ^
2025-07-04T13:19:01.5448900Z      
```

vvolkl commented:
Fixes https://github.com/cvmfs/cvmfs/issues/2792

christge-sft committed:
REVERT THIS: default enable the feature through the ctor

christge-sft committed:
[util] add a Unix Socket Cpp wrapper tailored to the needs of Quota-Cache mgr communication

christge-sft committed:
[util] support multiple clients

christge-sft committed:
[util] Complete redesign of the interface. See details bellow

Previously the interface was designed for exchanging strings
(std::string specifically).

To signal the end of transmission a specific sequence had to be read
during read.

Now any compact/contiguous type can be exchanged between the client and
the server.

In multiple exchanges it is the responsibility of the sender to inform
the receiver about the number of data, and of the receiver to poll for
the correct amount of data.

christge-sft committed:
[util] test hash exchanges

christge-sft committed:
[util] test for single server multiple clients

christge-sft committed:
[util] test with the realistic example

christge-sft committed:
[util] align with CVMFS logging

christge-sft committed:
[util] allow server to report the number of connected clients

christge-sft committed:
[util] allow asynchronous reading with try_read

christge-sft committed:
[util] make LocalUnixSocket inheritance-friendly and apply rule of 5

christge-sft committed:
[util] create a Quota and a Cache MGR endpoint. Allow a Quota mgr to collect hashes from all connected Cache mgrs

christge-sft committed:
[util] cleanup test suite

vvolkl commented on cvmfs/fd_refcount_mgr.h:
```diff
@@ -9,7 +9,7 @@
 #include "cache.h"
 #include "smallhash.h"
-
+#include <gtest/gtest_prod.h>
```
```suggestion
#include "duplex_testing.h"
```
See https://github.com/cvmfs/cvmfs/pull/3949

vvolkl commented on cvmfs/util/socket.h:
```diff
@@ -0,0 +1,320 @@
+template<ProcessType PT>
+class LocalUnixSocket {
+ public:
+  template<ProcessType X = PT,
```
This class needs some doxygen-style comments explaining  template args and args

christge-sft committed:
[util] create a CacheManagerSocket that can send a SmallHashDynamic map

christge-sft committed:
[util] tidy up codebase

christge-sft committed:
[util] close open sockets only

christge-sft committed:
[util] expose a release() API to the existing MakeSocket to allow the use of LocalUnixSocket within

christge-sft committed:
[util] retry binding after socket is unlinked; mimic MakeSocket

christge-sft committed:
[INTEGRATION] make cleanup policy selectible at runtime

christge-sft committed:
Revert "[INTEGRATION] get the information of open files from the Cache Manager"

This reverts commit 1ade8a58dd49cdfe4e05fd3abeef3fd5ae0b61b7.

christge-sft committed:
fixup! [util] create a CacheManagerSocket that can send a SmallHashDynamic map

christge-sft committed:
[util] Add corner case where zero hashes are transmitted

christge-sft committed:
fixup! [util] cleanup test suite

christge-sft committed:
fixup! fixup! [util] cleanup test suite

christge-sft committed:
fixup! fixup! fixup! [util] cleanup test suite

christge-sft committed:
[util] Fixes collect bugs

christge-sft committed:
[util] take into account non responsive cache managers

christge-sft committed:
[INTEGRATION] add to both manager the appropriate socket

christge-sft committed:
[INTEGRATION] Allow the QuotamManager to get the open files info from the dedicated socket

christge-sft committed:
[util] handle failure passing the responsibility entirely to the user

christge-sft committed:
fixup! [util] Complete redesign of the interface. See details bellow

christge-sft committed:
fixup! [util] take into account non responsive cache managers

christge-sft committed:
[INTEGRATION] WIP: add a magic attribute dedicated to open hashes

christge-sft committed:
[util] create a try connect socket functionality
